docs/recipes/UsageWithTypescript.md

```diff
@@ -6,259 +6,295 @@ hide_title: true
 
 ```ts
-// src/store/chat/reducers.ts
+import { AnyAction } from 'redux'
 
-import {
-  ChatState,
-  ChatActionTypes,
-  SEND_MESSAGE,
-  DELETE_MESSAGE
-} from './types'
+interface CounterState {
+  value: number
+}
 
-const initialState: ChatState = {
-  messages: []
+const initialState: CounterState = {
+  value: 0
 }
 
-export function chatReducer(
+export default function counterReducer(
   state = initialState,
-  action: ChatActionTypes
-): ChatState {
-  switch (action.type) {
-    case SEND_MESSAGE:
-      return {
-        messages: [...state.messages, action.payload]
-      }
-    case DELETE_MESSAGE:
-      return {
-        messages: state.messages.filter(
-          message => message.timestamp !== action.meta.timestamp
-        )
-      }
-    default:
-      return state
-  }
+  action: AnyAction<string>
+) {
+  // logic here
 }
 ```
 
-Type checked system reducer:
+However, the Redux core does export a `Reducer<State, Action>` type you can use as well.
 
-```ts
-// src/store/system/reducers.ts
+### Type Checking Middleware
 
-import { SystemState, SystemActionTypes, UPDATE_SESSION } from './types'
+[Middleware](../tutorials/fundamentals/part-4-store.md#middleware) are an extension mechanism for the Redux store. Middleware are composed into a pipeline that wrap the store's `dispatch` method, and have access to the store's `dispatch` and `getState` methods.
 
-const initialState: SystemState = {
-  loggedIn: false,
-  session: '',
-  userName: ''
-}
+The Redux core exports a `Middleware` type that can be used to correctly type a middleware function:
 
-export function systemReducer(
-  state = initialState,
-  action: SystemActionTypes
-): SystemState {
-  switch (action.type) {
-    case UPDATE_SESSION: {
-      return {
-        ...state,
-        ...action.payload
-      }
-    }
-    default:
-      return state
-  }
-}
+```ts
+export interface Middleware<
+  DispatchExt = {}, // legacy type parameter that is no longer relevant
+  S = any, // type of the Redux store state
+  D extends Dispatch = Dispatch // type of the dispatch method
+>
 ```
 
-We now need to generate the root reducer function, which is normally done using `combineReducers`. Note that we do not have to explicitly declare a new interface for RootState. We can use `ReturnType` to infer state shape from the `rootReducer`.
+A custom middleware should use the `Middleware` type, and pass the generic args for `S` (state) and `D` (dispatch) if needed:
 
 ```ts
-// src/store/index.ts
-
-import { systemReducer } from './system/reducers'
-import { chatReducer } from './chat/reducers'
+import { Middleware } from 'redux'
 
-const rootReducer = combineReducers({
-  system: systemReducer,
-  chat: chatReducer
-})
+import { RootState } from '../store'
 
-export type RootState = ReturnType<typeof rootReducer>
+export const exampleMiddleware: Middleware<
+  {}, // legacy type parameter added to satisfy interface signature
```

Comment:
As mentioned above, not legacy.
Also see https://github.com/reduxjs/redux-toolkit/blob/master/type-tests/files/configureStore.typetest.ts#L252-L264 as an example.